### PR TITLE
feat: add next document with build

### PR DESCRIPTION
**Pull request by clChenLiang**
# PR Details

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/runtime/plugin-runtime/src/cli/index.ts [hunk 1]
```diff
@@ -4,16 +4,18 @@ import type { CliPlugin } from '@modern-js/core';
...
   name: '@modern-js/runtime',
   post: [
     '@modern-js/plugin-router',
     '@modern-js/plugin-ssr',
+    '@modern-js/document',
     '@modern-js/plugin-state',
     '@modern-js/plugin-design-token',
   ],
-  usePlugins: [PluginState(), PluginSSR(), PluginRouter()],
+  usePlugins: [PluginState(), PluginSSR(), PluginRouter(), DocRouter()],
```
Comment (yimingjfe) [inline packages/runtime/plugin-runtime/src/cli/index.ts]: 为啥叫 DocRouter？

File: packages/runtime/plugin-runtime/src/document/cli.ts [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        });
+
+        const Document = (await import(htmlOutputFile)).default;
+
+        const HTMLElement = React.createElement(
+          DocumentContext.Provider,
+          { value: documentParams },
+          React.createElement(Document, null),
+        );
+        const html = ReactDomServer.renderToString(HTMLElement);
```
Comment (yimingjfe) [inline packages/runtime/plugin-runtime/src/document/cli.ts:128]: 可能用 renderToStaticMarkup 更合适？

File: tests/integration/app-document/src/App.css [hunk 3]
```diff
@@ -0,0 +1,119 @@
+html,
```
Comment (yimingjfe) [inline tests/integration/app-document/src/App.css:1]: 这里 css 好像没啥用的？

File: packages/runtime/plugin-runtime/tests/document/index.test.tsx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  Html,
+  Scripts,
+  Body,
+  Head,
+  Root,
+  DocumentContext,
+} from '../../src/document';
+import cliPlugin from '../../src/document/cli';
+
+describe.only('plugin-document', () => {
```
Comment (yimingjfe) [inline packages/runtime/plugin-runtime/tests/document/index.test.tsx:14]: 可以去掉  only 了

**Comment by clChenLiang** [comment]
1. document 的插件名, docume/cli/cli
2. 中文注释换成英文

File: packages/runtime/plugin-runtime/src/document/Body.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+} from './constants';
+import { DocumentStructrueContext } from './DocumentStructrueContext';
+import { DefaultRoot } from './Root';
+
+export function Body(props: { children?: any }) {
+  const { hasSetRoot } = useContext(DocumentStructrueContext);
+  const { children } = props;
+  return (
+    <body>
+      <noscript>{DOCUMENT_NO_SCRIPTE_PLACEHOLDER}</noscript>
```
Comment (zllkjc) [inline packages/runtime/plugin-runtime/src/document/Body.tsx:16]: noscript 移除掉吧。这个以后不放在模板里了

File: packages/runtime/plugin-runtime/src/document/Html.tsx [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+    // hasSetRoot,
+  ].every(item => item);
+
+  // todo: or throw an error
+  if (!notMissMustChild) {
+    return (
+      <html>
+        <body style={{ color: 'red' }}>
+          {`Miss the `}
+          {[
```
Comment (zllkjc) [inline packages/runtime/plugin-runtime/src/document/Html.tsx:68]: 这里是不是可以直接用 notMissMustChild 这个参数？

File: packages/runtime/plugin-runtime/src/document/Html.tsx [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            return item
+              ? null
+              : [
+                  'Head',
+                  'Body',
+                  // 'Scripts',
+                  // 'Root',
+                ][index];
+          })}
+          {` Element`}
```
Comment (zllkjc) [inline packages/runtime/plugin-runtime/src/document/Html.tsx:83]: 这是干嘛的

File: packages/runtime/plugin-runtime/src/document/cli/index.ts [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+          config: api.useConfigContext(),
+          entryName,
+          templateParameters,
+        });
+        const htmlOutputFile = path.join(
+          internalDirectory,
+          `./document/_${entryName}.html.js`,
+        );
+        // transform document file to html string
+        await build({
```
Comment (zllkjc) [inline packages/runtime/plugin-runtime/src/document/cli/index.ts:89]: 这里是 html 转 js 吗

File: packages/runtime/plugin-runtime/src/document/Html.tsx (hunk 6 above)
Comment (clChenLiang) [reply packages/runtime/plugin-runtime/src/document/Html.tsx:68]: notMissMustChild 是个 boolean 值。

File: packages/runtime/plugin-runtime/src/document/Html.tsx (hunk 7 above)
Comment (clChenLiang) [reply packages/runtime/plugin-runtime/src/document/Html.tsx:83]: 拼接 `Miss the xx Element` 文案

File: packages/runtime/plugin-runtime/src/document/Html.tsx (hunk 6 above)
Comment (clChenLiang) [reply packages/runtime/plugin-runtime/src/document/Html.tsx:68]: 这里是显示哪些必填的模块没有加上去

File: packages/runtime/plugin-runtime/src/document/cli/index.ts (hunk 8 above)
Comment (clChenLiang) [reply packages/runtime/plugin-runtime/src/document/cli/index.ts:88]: 是 document.tsx 转 js

File: packages/runtime/plugin-runtime/src/document/Body.tsx (hunk 5 above)
Comment (clChenLiang) [reply packages/runtime/plugin-runtime/src/document/Body.tsx]: done

**Review by zllkjc — APPROVED** [review]

**Pull request merged**